snapshots/devmapper/pool_device.go

```diff
@@ -118,35 +118,59 @@ func (p *PoolDevice) CreateThinDevice(ctx context.Context, deviceName string, vi
 		State: Unknown,
 	}
 
-	// Save initial device metadata and allocate new device ID from store
-	if err := p.metadata.AddDevice(ctx, info); err != nil {
-		return errors.Wrapf(err, "failed to save initial metadata for new thin device %q", deviceName)
-	}
+	var (
+		metaErr   error
+		devErr    error
+		activeErr error
+	)
 
 	defer func() {
-		if retErr == nil {
+		// We've created a devmapper device, but failed to activate it, try rollback everything
+		if activeErr != nil {
+			// Delete the device first.
+			delErr := p.deleteDevice(ctx, info)
+			if delErr != nil {
+				// Failed to rollback, mark the device as faulty and keep metadata in order to
+				// preserve the faulty device ID
+				retErr = multierror.Append(retErr, delErr, p.metadata.MarkFaulty(ctx, info))
+				return
+			}
+
+			// The devmapper device has been successfully deleted, deallocate device ID
+			if err := p.RemoveDevice(ctx, info.Name); err != nil {
+				retErr = multierror.Append(retErr, err)
+				return
+			}
+
 			return
 		}
 
-		// Rollback metadata
-		retErr = multierror.Append(retErr, p.metadata.RemoveDevice(ctx, info.Name))
+		// We're unable to create the devmapper device, most likely something wrong with the deviceID
+		if devErr != nil {
+			retErr = multierror.Append(retErr, p.metadata.MarkFaulty(ctx, info))
+			return
+		}
 	}()
 
-	// Create thin device
-	if err := p.createDevice(ctx, info); err != nil {
-		return err
+	// Save initial device metadata and allocate new device ID from store
+	metaErr = p.metadata.AddDevice(ctx, info)
+	if metaErr != nil {
+		return metaErr
 	}
 
-	defer func() {
-		if retErr == nil {
-			return
-		}
+	// Create thin device
+	devErr = p.createDevice(ctx, info)
+	if devErr != nil {
+		return devErr
+	}
 
-		// Rollback creation
-		retErr = multierror.Append(retErr, p.deleteDevice(ctx, info))
-	}()
+	// Activate thin device
+	activeErr = p.activateDevice(ctx, info)
+	if activeErr != nil {
+		return activeErr
+	}
 
```

Comment:
Hi,

What if crash happens in between these steps and rollback code doesn't get chance to run?

This possibility also applies to other places to change device or metadata, right?

Reply:
> What if crash happens in between these steps and rollback code doesn't get chance to run?

Each device mapper operation is wrapped with state transition function (have a look on [1](https://github.com/containerd/containerd/pull/3470/files#diff-50fc472e20d12e52c4fce854bd2daa96L94) and [2](https://github.com/containerd/containerd/pull/3470/files#diff-50fc472e20d12e52c4fce854bd2daa96R189)). If we're failed to activate a device and crash happens, so no `defer` was run, you'll end up with something like this in your metadata database (meaning that the device didn't finish the transition):
```
DevName=X ID=9 State=Activating Error="failed to activate"
```

Reply:
Essentially each operation with devmapper device is recorded (https://github.com/containerd/containerd/blob/master/snapshots/devmapper/device_info.go#L34), so you can get a clear picture about device state and last operation at any time. For example, for `CreateThinDevice` the following states will be used:
```
Unknown <-- preallocated metadata info and devID
Creating <-- createDevice
Created
Activating <-- activateDevice
Activated
```

Reply:
Hi,

thanks for your detailed explanation ;-) Yes, I know, in the crash case, the metadata DB can reflect correctly the device state before crashing, such as creating, activating etc.

Actually, my meaning is, in that case, there is no chance to the rollback code to mark the device as `faulty`, right?

If so, after machine comes up and containerd get started again, we still have "object already exists" error from the AddDevice() -  https://github.com/containerd/containerd/pull/3470/commits/878a3205cdfedf616f0db82381d6841ffdb9461c#diff-4132e94f99d36d21b09c446fb20687faL100.

Reply:
If you crash right in between create and rollback, then yes, a device won't be marked as `Faulty` one. In this case, what behavior would you expect to see?

Reply:
If a crash happens, containerd must restart again, so `devmapper` will be reloaded. Could we take the chance to scan the device list and mark the device as `faulty` who's state is not   in a final state - `Deactivated` and `Activated`. Or I feel it's safe to cleanup that device from metadata and thin pool completely.

Reply:
> If you crash right in between create and rollback, then yes, a device won't be marked as `Faulty` one. In this case, what behavior would you expect to see?

Hi，

I just come up this idea. I'd like to know what do you think？thanks.

diff --git a/snapshots/devmapper/pool_device.go b/snapshots/devmapper/pool_device.go
index 22e78c1..cf75b66 100644
--- a/snapshots/devmapper/pool_device.go
+++ b/snapshots/devmapper/pool_device.go
@@ -150,6 +150,15 @@ func (p *PoolDevice) CreateThinDevice(ctx context.Context, deviceName string, vi
                        retErr = multierror.Append(retErr, p.metadata.MarkFaulty(ctx, info.Name))
                        return
                }
+
+               // If ErrAlreadyExists returns, it indicates the device name is already in metadata,
+               // but not aware by upper layer. This may occur when crash happens and there is no
+               // chance to mark the device faulty. So, mark it here and upper layer will retry with
+               // the same device name.
+               if metaErr == ErrAlreadyExists {
+                       retErr = multierror.Append(retErr, p.metadata.MarkFaulty(ctx, info.Name))
+                       return
+               }

Reply:
I've addressed this use case in https://github.com/containerd/containerd/pull/3489. In case of crash, the snapshotter will mark devices as faulty after restart. Can you check if it works for you?

Reply:
> I've addressed this use case in #3489. In case of crash, the snapshotter will mark devices as faulty after restart. Can you check if it works for you?

Great, thanks!